prover/crates/bin/circuit_prover/src/main.rs

```diff
@@ -22,82 +21,105 @@ use zksync_prover_fri_types::PROVER_PROTOCOL_SEMANTIC_VERSION;
+    /// Corresponds to 1 CPU thread & ~2GB of RAM.
+    #[arg(short = 'l', long, default_value_t = 1)]
+    light_wvg_count: usize,
+    /// Number of heavy witness vector generators to run in parallel.
+    /// Corresponds to 1 CPU thread & ~9GB of RAM.
+    #[arg(short = 'h', long, default_value_t = 1)]
+    heavy_wvg_count: usize,
     /// Max VRAM to allocate. Useful if you want to limit the size of VRAM used.
     /// None corresponds to allocating all available VRAM.
-    #[arg(long)]
+    #[arg(short = 'm', long)]
     pub(crate) max_allocation: Option<usize>,
 }
 
 #[tokio::main]
 async fn main() -> anyhow::Result<()> {
-    let time = Instant::now();
+    let start_time = Instant::now();
     let opt = Cli::parse();
 
     let (observability_config, prover_config, object_store_config) = load_configs(opt.config_path)?;
-
     let _observability_guard = observability_config
         .install()
         .context("failed to install observability")?;
 
-    let wvg_count = opt.witness_vector_generator_count as u32;
-
-    let (connection_pool, object_store, setup_data_cache, hints) = load_resources(
+    let (connection_pool, object_store, prover_context, setup_data_cache, hints) = load_resources(
         opt.secrets_path,
+        opt.max_allocation,
         object_store_config,
         prover_config.setup_data_path.into(),
-        wvg_count,
     )
     .await
     .context("failed to load configs")?;
 
-    PROVER_BINARY_METRICS.start_up.observe(time.elapsed());
+    PROVER_BINARY_METRICS
+        .startup_time
+        .observe(start_time.elapsed());
 
     let cancellation_token = CancellationToken::new();
-    let backoff = Backoff::new(Duration::from_secs(5), Duration::from_secs(30));
 
     let mut tasks = vec![];
 
-    let (sender, receiver) = tokio::sync::mpsc::channel(5);
-
-    tracing::info!("Starting {wvg_count} Witness Vector Generators.");
-
-    for _ in 0..wvg_count {
-        let wvg = WitnessVectorGenerator::new(
-            object_store.clone(),
-            connection_pool.clone(),
-            PROVER_PROTOCOL_SEMANTIC_VERSION,
-            sender.clone(),
-            hints.clone(),
-        );
-        tasks.push(tokio::spawn(
-            wvg.run(cancellation_token.clone(), backoff.clone()),
-        ));
-    }
+    let (witness_vector_sender, witness_vector_receiver) = tokio::sync::mpsc::channel(CHANNEL_SIZE);
+
+    tracing::info!(
+        "Starting {} light WVGs and {} heavy WVGs.",
+        opt.light_wvg_count,
+        opt.heavy_wvg_count
+    );
+
+    let builder = WvgRunnerBuilder::new(
+        connection_pool.clone(),
+        object_store.clone(),
+        PROVER_PROTOCOL_SEMANTIC_VERSION,
+        hints.clone(),
+        witness_vector_sender,
+        cancellation_token.clone(),
+    );
+
+    let light_wvg_runner = builder.light_wvg_runner(opt.light_wvg_count);
+    let heavy_wvg_runner = builder.heavy_wvg_runner(opt.heavy_wvg_count);
```

Comment:
overall I like this approach with light/heavy separation. I understand we do that to prevent OOM by limiting the number of concurrent heavy jobs. 
Other option would be to associate the max potential RAM consumption with each job (so in a simple case we could say that Node are 9GB and all are 2GB), keep and adjust the current limit in memory as new tasks are loaded and limit spawning new tasks. This would in theory allow for more flexible resource utilization, but in practice it's doubtful it would actually matter and the current approach is simpler.